### PR TITLE
Implementation of fill

### PR DESCRIPTION
**Pull request by vucinicv**
This pr addresses the addition of the fill function to the lift library with a corresponding unit test.

### CODE REVIEW

File: lift/parallel.h [hunk 1]
```diff
@@ -350,6 +350,35 @@ struct parallel
                                            UniqueOutputIterator unique_keys_output,
                                            LengthOutputIterator run_lengths_output,
                                            allocation<system, uint8>& temp_storage);
+    //Fills an allocation with a value
```
Comment (nsubtil) [inline lift/parallel.h]: Remove this line.

File: lift/parallel.h [hunk 2]
```diff
@@ -350,6 +350,35 @@ struct parallel
...
+    /**
+     * Parallel for-each implementation. Sets each element in range [begin, end[ to value.
```
Comment (nsubtil) [inline lift/parallel.h]: Uh... this isn't for_each :)

File: lift/parallel.h [hunk 3]
```diff
@@ -350,6 +350,35 @@ struct parallel
...
+     *
+     * \tparam T                Data type for vector.
+     *
+     * \param vector            The memory region to fill with value
+     * \param value             Value to fill each element in  vector
+     * \param launch_parameters Same as in \ref fill
+     */
+    template <typename T>
+    static inline void fill(pointer<system, T>& vector,
+                                T value,
```
Comment (nsubtil) [inline lift/parallel.h]: Indentation.

File: lift/parallel.h [hunk 4]
```diff
@@ -350,6 +350,35 @@ struct parallel
...
+     * \param begin             Iterator pointing at the first element to be processed.
+     * \param end               Iterator pointing at the end of the range to be processed.
+     * \param value             The value with which to fill the range.
+     * \param launch_parameters Grid launch parameters for GPU backend.
+     */
+    template <typename InputIterator, typename T>
+    static inline void fill(InputIterator begin,
+                            InputIterator end,
+                            T value,
+                            int2 launch_parameters = { 0, 0 });
```
Comment (nsubtil) [inline lift/parallel.h]: It probably doesn't make sense to expose launch parameters for this. We should be able to figure out the optimal grid config since we know ahead of time what kernel we'll run.

File: lift/parallel/parallel_cuda.inl [hunk 5]
```diff
@@ -110,6 +110,40 @@ inline void parallel<cuda>::for_each(uint32 end,
...
+    LIFT_HOST_DEVICE void operator() (T &ref)
+    {
+        ref = value;
+    }
+};
+
+template <>
+template <typename InputIterator, typename T>
+inline void parallel<cuda>::fill(InputIterator begin,
+                            InputIterator end,
```
Comment (nsubtil) [inline lift/parallel/parallel_cuda.inl]: Indentation

File: lift/parallel/parallel_cuda.inl [hunk 6]
```diff
@@ -110,6 +110,40 @@ inline void parallel<cuda>::for_each(uint32 end,
...
+                            T value,
+                            int2 launch_parameters)
+{
+    for_each(begin, end, fill_by_reference<T>(value), launch_parameters);
+}
+
+template <>
+template <typename T>
+inline void parallel<cuda>::fill(pointer<cuda, T>& vector,
+                            T value,
```
Comment (nsubtil) [inline lift/parallel/parallel_cuda.inl]: Indentation

File: lift/parallel/parallel_host.inl [hunk 7]
```diff
@@ -95,6 +95,25 @@ inline void parallel<host>::for_each(uint32 end,
 }
 
 template <>
+template <typename InputIterator, typename T>
+inline void parallel<host>::fill(InputIterator begin,
+                            InputIterator end,
```
Comment (nsubtil) [inline lift/parallel/parallel_host.inl]: Indentation

File: lift/parallel/parallel_host.inl [hunk 8]
```diff
@@ -95,6 +95,25 @@ inline void parallel<host>::for_each(uint32 end,
...
+                            T value,
+                            int2 launch_parameters)
+{
+    for_each(begin, end, fill_by_reference<T>(value), launch_parameters);
+}
+
+template <>
+template <typename T>
+inline void parallel<host>::fill(pointer<host, T>& vector,
+                            T value,
```
Comment (nsubtil) [inline lift/parallel/parallel_host.inl]: Indentation

File: tests/fill.cu [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    bool test_success = true;
+    uint32 size = 1000;
+    int fill_val = 10;
+
+    allocation<system, int> data(size);
+    parallel<system>::fill(data, fill_val);
+
+    for (int i = 0; i < size; i++)
+    {
+        if (data.peek(i) != fill_val)
```
Comment (nsubtil) [inline tests/fill.cu]: This should be LIFT_TEST_CHECK(data.peek(i) == fill_val). And remove test_success :)

File: tests/fill.cu [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        {
+            test_success = false;
+        }
+    }  
+    LIFT_TEST_CHECK(test_success == true);
+}
+LIFT_TEST_FUNC(fill_input_iter_test, fill_vector);
```
Comment (nsubtil) [inline tests/fill.cu]: Passing the same test name twice is not going to work. This is actually something that the test harness should check for.

File: tests/fill.cu (hunk 9 above)
Comment (vucinicv) [inline tests/fill.cu]: Hey so I just want to run this by you, when it is in this format, the error output is one line long. If I were to remove test_success and run the check in the for loop, a failure would print out ~1000 lines to stdout. Using the current method, the output log would be cleaner.

File: tests/fill.cu (hunk 10 above)
Comment (vucinicv) [inline tests/fill.cu]: Opened an issue to address this, will be addressed in a separate pr.

File: tests/fill.cu (hunk 9 above)
Comment (nsubtil) [inline tests/fill.cu]: That is the intent. That way, test logs will show if there's a single element failure or a widespread failure, which is useful data to have.

File: tests/fill.cu (hunk 10 above)
Comment (nsubtil) [inline tests/fill.cu]: I saw that, thanks!

Note that we still need to fix the duplicate test name in this code, though.

File: tests/fill.cu (hunk 9 above)
Comment (vucinicv) [inline tests/fill.cu]: :+1: got it, changing it now

File: tests/fill.cu (hunk 10 above)
Comment (vucinicv) [inline tests/fill.cu]: Hey so as far as I can tell, in the most recent code LIFT_TEST_FUNC will take the test name and split it into  test_name_cuda and test_name_host, so although I only make one call LIFT_TEST_FUNC(fill_input_iter_test, fill_vector); it should make both tests?
